File: Toolboxes/Tisean_3.0.1/unsupported/rank.c
/* box assisted sorting/ranking utilities    */
/* Copyright (C) T. Schreiber (1998)         */

#include <stdlib.h>
#include <stdio.h>
#define MAX(a,b) ((a)>(b)? (a) : (b))
#define MIN(a,b) ((a)<(b)? (a) : (b))
#define EMPTY (-1)

void minmax(long nmax, double *x, double *xmin, double *xmax)
{
   long i;

   for(*xmin=x[0], *xmax=x[0], i=0; i<nmax; i++){
       *xmax=MAX(*xmax,x[i]);
       *xmin=MIN(*xmin,x[i]);
   }
}

void rank(long nmax, double *x, long *list)
{
   long *jptr, nl, n, i, ip, ipp;
   double xmin, xmax, sc, xn;

   nl=nmax/2;
   jptr=(long *)malloc(nl*sizeof(long));
   for(n=0; n<nl; n++) 
      jptr[n]=EMPTY;
   minmax(nmax, x, &xmin, &xmax);
   if(xmax==xmin){
      for(n=0; n<nmax; n++)
         list[n]=n;
      return;
   }
   sc=(nl-1)/(xmax-xmin);

   for(n=0; n<nmax; n++){
      xn=x[n];
      i=(long)((xn-xmin)*sc);
      ip=jptr[i];
      if( (ip==EMPTY) || (xn<=x[ip]) ) {
         jptr[i]=n;
      } else {
	do{
           ipp=ip;
	   ip=list[ip];
        }
	while( (ip!=EMPTY) && (xn>x[ip]) );
	list[ipp]=n;
      }
      list[n]=ip;
   }

   for(n=0, i=0; i<nl; i++){
      ip=jptr[i];
      while(ip!=EMPTY){
         ipp=ip;
         ip=list[ip];
         list[ipp]=++n;
      }
   }

   for(n=0; n<nmax; n++) list[n]-=1;
   free(jptr);
}

void rank2index(long nmax, long *list)
{
   long n, ib, im, it;

   for(n=0; n<nmax; n++)
      list[n]=-list[n];

   for(n=0; n<nmax; n++){
      if(list[n]<0){
         ib=n;
         im=-list[n];
         while(1){
            it=-list[im];
            list[im]=ib;
	    if(it==n){
               list[n]=im;
               break;
            } 
            ib=im;
            im=it;
         }
      }
   }
}

void indexx(long nmax, double *x, long *list)
{
   rank(nmax,x,list);
   rank2index(nmax,list);
}

void rank2sort(long nmax, double *x, long *list)
{
   long n, ib, it;
   double hb, ht;

   for(n=0; n<nmax; n++)
      list[n]=-list[n];

   for(n=0; n<nmax; n++){
      if(list[n]<0){
         ib=n;
         hb=x[n];
         while(1){
            it=-list[ib];
            list[ib]=it;
	    ht=x[it];
            x[it]=hb;
	    if(it==n) break;
            ib=it;
            hb=ht;
          }
       }
   }
}

void sort(long nmax, double *x)
{
   long *list;

   list=(long *)malloc(nmax*sizeof(long));
   rank(nmax,x,list);
   rank2sort(nmax,x,list);
   free(list);
}

File: Toolboxes/Tisean_3.0.1/source_f/rank.f
c===========================================================================
c
c   This file is part of TISEAN
c 
c   Copyright (c) 1998-2007 Rainer Hegger, Holger Kantz, Thomas Schreiber
c 
c   TISEAN is free software; you can redistribute it and/or modify
c   it under the terms of the GNU General Public License as published by
c   the Free Software Foundation; either version 2 of the License, or
c   (at your option) any later version.
c
c   TISEAN is distributed in the hope that it will be useful,
c   but WITHOUT ANY WARRANTY; without even the implied warranty of
c   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
c   GNU General Public License for more details.
c
c   You should have received a copy of the GNU General Public License
c   along with TISEAN; if not, write to the Free Software
c   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
c
c===========================================================================
c   box assisted sorting/ranking utilities 
c   author T. Schreiber (1998) based on earlier versions
c===========================================================================
      subroutine rank(nmax,x,list)
c  rank points in x
      parameter(nptr=100000)
      dimension x(nmax), list(nmax), jptr(0:nptr)

      call minmax(nmax,x,xmin,xmax)
      if(xmin.eq.xmax) then
         do 10 n=1,nmax
 10         list(n)=n
         return
      endif
      nl=min(nptr,nmax/2)
      sc=(nl-1)/(xmax-xmin)
      do 20 i=0,nl
 20      jptr(i)=0
      do 30 n=1,nmax
         xn=x(n)
         i=int((xn-xmin)*sc)
         ip=jptr(i)
         if ((ip.eq.0).or.(xn.le.x(ip))) then
            jptr(i)=n
         else
 1          ipp=ip
            ip=list(ip)
            if ((ip.gt.0).and.(xn.gt.x(ip))) goto 1
            list(ipp)=n
         endif
 30      list(n)=ip
      n=0
      do 40 i=0,nl
         ip=jptr(i)
 2       if (ip.eq.0) goto 40
         n=n+1
         ipp=ip
         ip=list(ip)
         list(ipp)=n
         goto 2
40       continue
      end

      subroutine indexx(nmax,x,list)
c make index table using rank
      dimension x(nmax), list(nmax)
      
      call rank(nmax,x,list)
      call rank2index(nmax,list)
      end

      subroutine rank2index(nmax,list)
c converts a list of ranks into an index table (or vice versa) in place
      integer list(nmax)

      do 10 n=1,nmax
 10      list(n)=-list(n)
      do 20 n=1,nmax
         if(list(n).gt.0) goto 20               ! has been put in place already
         ib=n
         im=-list(n)
 1       it=-list(im)
         list(im)=ib
         if(it.ne.n) then
            ib=im
            im=it
            goto 1
         else
            list(n)=im
         endif
 20      continue
      end

      subroutine sort(nmax,x,list)
c sort using rank and rank2sort
      dimension x(nmax), list(nmax)

      call rank(nmax,x,list)
      call rank2sort(nmax,x,list)
      end

      subroutine rank2sort(nmax,x,list)
c sort x using list of ranks
      dimension x(nmax), list(nmax)

      do 10 n=1,nmax
 10      list(n)=-list(n)
      do 20 n=1,nmax
         if(list(n).gt.0) goto 20               ! has been put in place already
         ib=n
         hb=x(n)
 1       it=-list(ib)
         list(ib)=it
         ht=x(it)
         x(it)=hb
         if(it.ne.n) then
            ib=it
            hb=ht
            goto 1
         endif
 20      continue
      end

      subroutine index2sort(nmax,x,list)
c sort x using list of indices
      dimension x(nmax), list(nmax)

      do 10 n=1,nmax
 10      list(n)=-list(n)
      do 20 n=1,nmax
         if(list(n).gt.0) goto 20               ! has been put in place already
         ib=n
         h=x(n)
 1       it=-list(ib)
         list(ib)=it
         if(it.ne.n) then
            x(ib)=x(it)
            ib=it
            goto 1
         else
            x(ib)=h
         endif
 20      continue
      end

      function which(nmax,x,k,list)
      dimension x(nmax), list(nmax)

      call indexx(nmax,x,list)
      which=x(list(k))
      end

